internal/tokencache/default_storage_manager.go

```diff
@@ -84,46 +89,57 @@ func (m *defaultStorageManager) WriteAccessToken(accessToken *accessTokenCacheIt
 	return nil
 }
 
-func (m *defaultStorageManager) ReadRefreshToken(homeID string, envAliases []string, familyID, clientID string) *refreshTokenCacheItem {
+func (m *defaultStorageManager) ReadRefreshToken(homeID string, envAliases []string, familyID, clientID string) (refreshTokenCacheItem, error) {
 	lock.RLock()
 	defer lock.RUnlock()
 
-	if familyID != "" {
-		for _, rt := range m.refreshTokens {
-			if matchFamilyRefreshToken(rt, homeID, envAliases) {
-				return rt
-			}
-		}
-		for _, rt := range m.refreshTokens {
-			if matchClientIDRefreshToken(rt, homeID, envAliases, clientID) {
-				return rt
-			}
+	byFamily := func(rt refreshTokenCacheItem) bool {
+		return matchFamilyRefreshToken(rt, homeID, envAliases)
+	}
+	byClient := func(rt refreshTokenCacheItem) bool {
+		return matchClientIDRefreshToken(rt, homeID, envAliases, clientID)
+	}
+
+	var matchers []func(rt refreshTokenCacheItem) bool
+	if familyID == "" {
+		matchers = []func(rt refreshTokenCacheItem) bool{
+			byClient, byFamily,
 		}
 	} else {
-		for _, rt := range m.refreshTokens {
-			if matchClientIDRefreshToken(rt, homeID, envAliases, clientID) {
-				return rt
-			}
+		matchers = []func(rt refreshTokenCacheItem) bool{
+			byFamily, byClient,
 		}
+	}
 
+	// TODO(jdoak): So, I redid this code to make it clearer that we
+	// changed order based on if FamilyID was set. But I'm not sure
+	// why I have to loop twice over the tokens instead of just
+	// trying both matchers.  If someone could shed some light on this
+	// I could make a note for future maintainers. If order doesn't
+	// matter, then we could ditch byFamily and byClient for cleaner
+	// code.
+	for _, matcher := range matchers {
```

Comment:
@bgavrilMS can you review this part? thanks.

Reply:
ping @bgavrilMS

Reply:
The algorithm is a bit more complex than this, I assume there are some tests covering everything. I would keep the order as is. 

The algorithm is 

1. If application is NOT part of the family, search by client_ID
2. If app is part of the family or if we DO NOT KNOW if it's part of the family, search by family ID, then by client_id (we will know if an app is part of the family after the first token response). 

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/311fe8b16e7c293462806f397e189a6aa1159769/src/client/Microsoft.Identity.Client/Internal/Requests/Silent/CacheSilentStrategy.cs#L95

Reply:
@bgavrilMS 

There is a test, please verify it looks correct (as it currently passes) and advise if we should add another test:
internal/tokencache/default_storage_manager_test.go: TestDefaultStorageManagerReadRefreshToken

Reply:
Wow, such a verbose test. I think this is fine. Closing.

Reply:
Opened a bug for you to identify a test that we should add that will fail due to current method.  Will update the code after adding said test.